packages/react-aria-components/test/AriaMenu.test-util.tsx

```diff
@@ -0,0 +1,742 @@
+import {
+  pointerMap
+} from '@react-spectrum/test-utils-internal';
+import {User} from '@react-aria/test-utils';
+import userEvent from '@testing-library/user-event';
+
+let describeInteractions = (name, tests) => describe.each`
+  interactionType
+  ${'mouse'}
+  ${'keyboard'}
+  ${'touch'}
+`(`${name} - $interactionType`, tests);
+
+let triggerText = 'Menu Button';
+
+let ariaRender = render;
+interface AriaBaseTestProps {
+  setup?: () => void,
+  prefix?: string
+}
+interface AriaMenuTestProps extends AriaBaseTestProps {
+  renderers: {
+    // needs at least three child items, all enabled
+    standard: (props?: {name: string}) => ReturnType<typeof ariaRender>,
+    // trigger must be disabled
+    disabledTrigger?: (props?: {name: string}) => ReturnType<typeof ariaRender>,
+    // needs at least three child items, all enabled, with single selection
+    singleSelection?: (props?: {name: string}) => ReturnType<typeof ariaRender>,
+    // needs at least four child items, all enabled, with single selection
+    multipleSelection?: (props?: {name: string}) => ReturnType<typeof ariaRender>,
+    // needs a focusable sibling after the trigger with the label 'after'
+    // TODO: better to have tests return JSX and I call `render`? could allow me to inject elements in the DOM more easily
+    siblingFocusableElement?: (props?: {name: string}) => ReturnType<typeof ariaRender>,
+    // needs two menus that are siblings
+    multipleMenus?: (props?: {name: string}) => ReturnType<typeof ariaRender>,
+    // Menu should only open on long press
+    longPress?: (props?: {name: string}) => ReturnType<typeof ariaRender>,
+    // Menu must have two levels of submenus
+    submenus?: (props?: {name: string}) => ReturnType<typeof ariaRender>
+  }
+}
+export const AriaMenuTests = ({renderers, setup, prefix}: AriaMenuTestProps) => {
+  describe(prefix ? prefix + 'AriaMenuTrigger' : 'AriaMenuTrigger', function () {
+    let onOpenChange = jest.fn();
+    let onOpen = jest.fn();
+    let onClose = jest.fn();
+    let onSelect = jest.fn();
+    let onSelectionChange = jest.fn();
+    let user;
+    let testUtilUser = new User();
+    setup?.();
+
+    beforeAll(function () {
+      user = userEvent.setup({delay: null, pointerMap});
+      window.HTMLElement.prototype.scrollIntoView = jest.fn();
+      jest.useFakeTimers();
+    });
+
+    afterEach(() => {
+      onOpenChange.mockClear();
+      onOpen.mockClear();
+      onClose.mockClear();
+      onSelect.mockClear();
+      onSelectionChange.mockClear();
+      act(() => jest.runAllTimers());
+    });
+
+
+  // it('should support aria-label on the menu items', () => {
+  //   let {getAllByRole} = renderMenu({}, {'aria-label': 'test'});
+
+  //   for (let menuitem of getAllByRole('menuitem')) {
+  //     expect(menuitem).toHaveAttribute('aria-label', 'test');
+  //   }
+  // });
+
+    it('has default behavior (button renders, menu is closed)', function () {
+      let tree = renderers.standard();
```

Comment:
Interesting, I like the reusability of these tests, essentially establishes a standard suite that theoretically ANY implementation of the aria menu pattern could directly use. I assume you brain stormed these first and then pulled out/reused duplicates from the existing suites?

Reply:
I did it a little more organically than that, I pulled out from the current React Spectrum MenuTrigger almost 1:1, then started grouping based on how different the render requirements were.

I set out to have the tests be agnostic of dynamic case, or static case, or uncontrolled, or controlled, but that was as much as I had originally planned ahead